### PR TITLE
teamcity: add script to run PGO build and collect profiles

### PR DESCRIPTION
**Pull request by rickystewart**
This script ties together the pieces of the workflow: start a TC job, wait for it to finish, downloads the artifacts, parses the profiles, merges them, and produces a final result.

Part of: CRDB-44692
Epic: CRDB-41952
Release note: None

### CODE REVIEW

File: pkg/cmd/run-pgo-build/main.go [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+	for {
+		if build.FinishDate != "" {
+			break
+		}
+		time.Sleep(3 * time.Minute)
+		lookupID := build.ID
+		newBuild, err := getBuild(lookupID)
+		if err != nil {
+			fmt.Printf("failed to get build %d; got error %+v\n", lookupID, err)
```
Comment (rail) [inline pkg/cmd/run-pgo-build/main.go:232]: A nit. Can you add some wording that this is expected and the operation will be retried.

File: pkg/cmd/run-pgo-build/main.go [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+		}
+		build = newBuild
+	}
+
+	if build.Status != "SUCCESS" {
+		panic(fmt.Sprintf("expected build to succeed; got status %s for build %+v", build.Status, build))
```
Comment (rail) [inline pkg/cmd/run-pgo-build/main.go:238]: Maybe use `log.Fatal()` instead - I don't think we care about the trace backs...

File: pkg/cmd/run-pgo-build/main.go [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+	io.ReadCloser
+}
+
+func doRequest(req *http.Request) (*Build, error) {
+	req.SetBasicAuth(username, password)
+	req.Header.Add("Accept", "application/json")
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
```
Comment (rail) [inline pkg/cmd/run-pgo-build/main.go:59]: Do we care about the status code? Not sure if you want to proceed for statusCode >= 400.

File: pkg/cmd/run-pgo-build/main.go [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+func downloadArtifacts(buildID int64, tmpDir string) (ReadAtCloser, int64, error) {
+	req, err := http.NewRequest("GET", fmt.Sprintf("https://%s/repository/downloadAll/%s/%d:id", serverURL, buildConfigID, buildID), nil)
+	if err != nil {
+		return nil, 0, err
+	}
+	req.SetBasicAuth(username, password)
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, 0, err
+	}
```
Comment (rail) [inline pkg/cmd/run-pgo-build/main.go:112]: Same here. What if TC for some reason returns a 404?

**Review by rail — APPROVED** [review]

File: pkg/cmd/run-pgo-build/main.go (hunk 2 above)
Comment (rickystewart) [reply pkg/cmd/run-pgo-build/main.go:238]: On the contrary I've already found the tracebacks helpful while debugging this :)

**Comment by rickystewart** [comment]
TFTR! bors r=rail

**Comment by rickystewart** [comment]
bors r=rail

**Pull request merged**